Intergrate roofline benchmark into rocprof-compute

benrichard-amd opened the pull request:
## Motivation

Currently the roofline benchmarks live inside a separate binary rocm-amdgpu-bench, which is called by rocprof-compute. This PR brings the roofline benchmarks into the rocprofiler-compute. See https://github.com/ROCm/rocm-amdgpu-bench/.

## Technical Details
The GPU kernels are moved into the rocprofiler-compute project and compiled at runtime using hipRTC. Benchmark results are recorded in a CSV file. From the user perspective, nothing has changed. Everything works as before.

### HIP and HIPRTC wrappers

`src/hip.py`
`src/hiprtc.py`

Lightweight wrappers were created about HIP and HIPRTC runtime libraries. The method map directly to the hip and hiprtc functions, with some Python comforts thrown in e.g. objects are garbage collected instead of requiring explicit destruction.

### Benchmark

`src/utils/benchmark.py`

The benchmark code lives here. I've tried to minimize code duplication, so typically for each test type (e.g. VALU FLOPS or MFMA FLOPS) there will be a function where the type is paramterized e.g. `mfma_bench(device, type, unit, rate)`.


To make testing and debugging easier, `benchmark.py` contains an entry point so that the benchmark can be called directly from the commandline e.g. `python src/utils/benchmark.py`

### Changes from rocm-amdgpu-bench
Use correct numbers for MFMA I8 and BF16  (See https://github.com/ROCm/rocm-amdgpu-bench/issues/24)

## Test Plan
* Verify metrics are similar to `rocm-amdgpu-bench`.
* Verify CSV columns have same names
* Verify similar `rocprof-compute analyze` output, in particular roofline section


## Test Result



## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

benrichard-amd committed:
Python roofline benchmark initial commit

benrichard-amd committed:
Add benchmark.py

benrichard-amd committed:
Add progress bar to benchmarks

benrichard-amd committed:
Remove some ifdefs some benchmark kernel code

benrichard-amd committed:
Fix crash on gfx908

-Separate the MFMA FP32 code from the others so that it can build on gfx908
-Added main function so benchmark can be called without rocprof-compute, useful for testing

benrichard-amd committed:
Add copyright headers

benrichard-amd committed:
Remove commented-out code

benrichard-amd committed:
Remove commented-out functions

benrichard-amd committed:
Fix typo in MFMA F16 CSV header

benrichard-amd committed:
Display GPU HIP ID and gfx arch

feizheng10 commented on projects/rocprofiler-compute/src/hip/hip.py:
```diff
@@ -0,0 +1,457 @@
+##############################################################################
+# MIT License
+#
+# Copyright (c) 2021 - 2025 Advanced Micro Devices, Inc. All Rights Reserved.
```
If it is a new file, maybe keep 2025 only?

benrichard-amd replied:
Fixed.

feizheng10 reviewed (approved):
Cool!!!

vedithal-amd commented on projects/rocprofiler-compute/src/utils/benchmark.py:
```diff
@@ -0,0 +1,1104 @@
+##############################################################################
+# MIT License
+#
+# Copyright (c) 2021 - 2025 Advanced Micro Devices, Inc. All Rights Reserved.
```
As far as I remember, for new files we just say the year of introduction which is 2025 here

benrichard-amd replied:
Fixed.

vedithal-amd commented on projects/rocprofiler-compute/src/hip/hip.py:
```diff
@@ -0,0 +1,457 @@
+##############################################################################
```
We should specify for which ROCm version these HIP bindings correspond to.
As HIP API changes with ROCm release, these python bindings also need to be updated.
So, tracking ROCm version with python bindings is useful for debugging.

On a separate note, what happens if someone using HIP from ROCm 6.4 tries to run the benchmarks? If it is going to fail, maybe it is a good idea to keep the roofline binaries for backward compatibility purposes.

benrichard-amd replied:
> We should specify for which ROCm version these HIP bindings correspond to.
As HIP API changes with ROCm release, these python bindings also need to be updated.
So, tracking ROCm version with python bindings is useful for debugging.

It's not a huge concern right now. The functions exposed are the most basic HIP functions that have existed forever and not likely to change. Should be mindful when adding more, though.

But to answer your question, it corresponds to HIP 6.0.0 since that's what version of `struct hipDeviceProp_t` is mirrored. (There was an ABI break in 5.0 -> 6.0, and the struct hasn't changed for 7.0). If there are ABI changes in the future we might need to add some ROCm version detection and create some codepaths for each.

> On a separate note, what happens if someone using HIP from ROCm 6.4 tries to run the benchmarks? If it is going to fail, maybe it is a good idea to keep the roofline binaries for backward compatibility purposes.

It will pick up whatever `libamdhip64.so` and `libhiprtc.so` are in the library path. So if they are on HIP 6.4, it will pick up the 6.4 libraries, which are compatible with the bindings we have.

vedithal-amd replied:
Good to know, just wanted to highlight the potential need of backward compatible bindings in the future.

benrichard-amd committed:
Update year on copyright headers

benrichard-amd committed:
Multiple fixes for gfx90a

-Fix L2 bw test size
-Add missing L1 bw test
-Separate out MFMA F16/BF16/F64/I8 source since F8/F6/F4 is not supprted on gfx90a

benrichard-amd committed:
Update comments

benrichard-amd committed:
Fix formatting

feizheng10 committed:
Merge branch 'develop' into benrichard-amd/roofline-python

benrichard-amd committed:
Merge branch 'develop' into benrichard-amd/roofline-python

benrichard-amd committed:
Add type annotations

Fixes warnings generated by ruff linter.

benrichard-amd committed:
Fix formatting

benrichard-amd committed:
Remove mibench/roofline unit tests

These tests depend on the roofline binaries, which do not exist anymore.

benrichard-amd committed:
Remove unused imports

benrichard-amd committed:
Merge branch 'develop' into benrichard-amd/roofline-python

cfallows-amd commented on projects/rocprofiler-compute/src/utils/benchmark.py:
```diff
@@ -0,0 +1,1142 @@
+* 3 = bf6
+* 4 = fp4
+*/
+template<int datatype> __global__ void mfma_f8f6f4(int iter, float *dummy)
```
~We need to provide 'datatype' arg for line 834. This builtin is a new style for MI350+ that incorporates new MX datatype formatting, CBSZ and BLGP args specify matrix A and B format respectively in the inputs.
(unfortunately) we need to separate the calls for f4 and f6 by providing 'datatype' with iter and dummy at line 925. The other cases below were added preemptively assuming we would at some point be asked to support the data types.~

Ben has pointed out template location at line 100- negates above, functional as intended. Resolving comment.

cfallows-amd commented on projects/rocprofiler-compute/src/utils/benchmark.py:
```diff
@@ -0,0 +1,1142 @@
+    }
+}
+
+extern "C" __global__ void mfma_f8(int iter, float *dummy)
```
We need to keep F8 separate for MI300 that does not support the f8f6f4 MX datatype mix mode builtins. We can use this for MI350+ as well (easier to default to this for any series that supports f8), but we could utilize f8f6f4 above for F8xF8 case for MI350+.

Edit: Ben has pointed out template location at line 100- negates above comment as a "necessary" separation as it is functioning as intended, but keeping comment as optional for clarity in separating F8 MFMA instruction from F8F6F4 MX MFMA instructions.

cfallows-amd commented on projects/rocprofiler-compute/src/utils/benchmark.py:
```diff
@@ -0,0 +1,1142 @@
+using f32_16vec = __attribute__((__vector_size__(16 * sizeof(float)))) float;
+using f16_2vec = __attribute__((__vector_size__(2 * sizeof(__2f16))))  float;
+
+/* Datatypes available for scale mfma f8f6f4 builtin
```
Possible extra comment content if we want to add the formats to the corresponding MX formats:
0 FP8 E4M3
1 FP8 E5M2
2 FP6 E2M3
3 FP6 E3M2
4 FP4 E2M1

Just noting source of identifier<->format table is the CDNA4 ISA: https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-cdna4-instruction-set-architecture.pdf

benrichard-amd replied:
I defined these as constants instead.

cfallows-amd commented:
Quick note- RHEL8/9 workflow fix has been committed 12/1, build fail unrelated to this PR.

cfallows-amd reviewed (changes requested):
F8F6F4 needs to provide extra MX data type selection argument, and separation from F8. See previous comments for specific lines. Can provide more detail if reasoning or the former implementation is not clear.

benrichard-amd committed:
Put the F8 kernel into its own module
Use preprocessor consants for the MX types instead of numbers

benrichard-amd committed:
Fix constant name FP6_E3M2 -> BF6_E3M2

benrichard-amd committed:
Merge branch 'develop' into benrichard-amd/roofline-python

benrichard-amd committed:
Update constant name FP8_E5M2 -> BF8_E5M2 to be consistent with ISA documentation

benrichard-amd committed:
Merge branch 'benrichard-amd/roofline-python' of github.com:rocm/rocm-systems into benrichard-amd/roofline-python

cfallows-amd commented:
/AzurePipelines run rocm-ci-caller
